pyriemann_qiskit/classification.py

```diff
@@ -743,3 +745,210 @@ def predict(self, X):
+        self.classes_ = np.unique(y)
+        print("Classes:", self.classes_)
+        print("Class 1", sum(y))
+        print("Class 2", len(y) - sum(y))
+        if max(self.classes_) > 1:
+            raise Exception("Currently designed for two classes 0 et 1")
+
+        for c in self.classes_:
+            self.matrices_per_class_[c] = []
+
+        for i in range(0, len(y)):
+            self.matrices_per_class_[y[i]].append(X[i, :, :])
+
+        for c in self.classes_:
+            self.matrices_per_class_[c] = np.array(self.matrices_per_class_[c])
+
+        print("End NCH Train")
+
+    def _process_sample(self, test_sample):
+        best_distance = -1
+        best_class = -1
+
+        for c in self.classes_:
+            distance = qdistance_logeuclid_to_convex_hull(
+                self.matrices_per_class_[c], test_sample
+            )
+            # print("Distance: ", distance)
+
+            if best_distance == -1:
+                best_distance = distance
+                best_class = c
+            elif distance < best_distance:
+                best_distance = distance
+                best_class = c
+
+        return best_class
+
+    def predict(self, X):
+        """Calculates the predictions.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            Input vector, where `n_samples` is the number of samples and
+            `n_features` is the number of features.
+
+        Returns
+        -------
+        pred : array, shape (n_samples,)
+            Class labels for samples in X.
+        """
+
+        print("Start NCH Predict")
+        pred = []
+
+        # testing only!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+        # labels = [1] * X.shape[0]
+        # print("End NCH Predict")
+        # return labels
+
+        print("Total test samples:", X.shape[0])
+        k = 0
+
+        parallel = True
+
+        if parallel:
+            pred = Parallel(n_jobs=12)(
+                delayed(self._process_sample)(test_sample) for test_sample in X
+            )
+
+        else:
+            for test_sample in X:
+                k = k + 1
+                print("Processing sample:", k, " / ", X.shape[0])
+
+                best_distance = -1
+                best_class = -1
+
+                for c in self.classes_:
+                    distance = qdistance_logeuclid_to_convex_hull(
+                        self.matrices_per_class_[c], test_sample
+                    )
+                    print("Distance: ", distance)
+
+                    if best_distance == -1:
+                        best_distance = distance
+                        best_class = c
+                    elif distance < best_distance:
+                        best_distance = distance
+                        best_class = c
+
+                pred.append(best_class)
+                print("Predicted: ", best_class)
+
+        print("End NCH Predict")
+        return np.array(pred)
```

Comment:
```suggestion
    def __init__(self, n_jobs=1):
        """Init."""
        self.n_jobs = n_jobs

    def fit(self, X, y):
        """Fit (store the training data).

        Parameters
        ----------
        X : ndarray, shape (n_matrices, n_channels, n_channels)
            Set of SPD matrices.
        y : ndarray, shape (n_matrices,)
            Labels for each matrix.
        sample_weight : None
            Not used, here for compatibility with sklearn API.

        Returns
        -------
        self : NearestConvexHull instance
            The NearestConvexHull instance.
        """

        self.mats_ = X
        self.labels_ = y
        self.classes_ = np.unique(y)

    def _predict_distances(self, X):
        """Helper to predict the distance. Equivalent to transform."""
        dist = [
            qdistance_logeuclid_to_convex_hull(
                self.mats_[self.labels_ == ll],
                X
            )
            for ll in self.classes_
        ]
        dist = np.concatenate(dist, axis=1)
        return dist

    def predict(self, X):
        """Get the predictions.

        Parameters
        ----------
        X : ndarray, shape (n_matrices, n_channels, n_channels)
            Set of SPD matrices.

        Returns
        -------
        pred : ndarray of int, shape (n_matrices,)
            Predictions for each matrix according to the closest convex hull.
        """
        dist = self._predict_distances(X)
        return self.classes_[dist.argmin(axis=1)]

    def transform(self, X):
        """Get the distance to each convex hull.

        Parameters
        ----------
        X : ndarray, shape (n_matrices, n_channels, n_channels)
            Set of SPD matrices.

        Returns
        -------
        dist : ndarray, shape (n_matrices, n_classes)
            The distance to each convex hull.
        """
        return self._predict_distances(X)
```

Reply:
I would like to keep my implementation for while except if there is an error in the implementation. It allows me to better test and debug.
We can switch to this implementation for the final commit. I see that your implementation is similar to the ones already in pyRiemann.

Reply:
qdistance_logeuclid_to_convex_hull has a different signature in the second parameter than in pyriemann so it could not be direct replacement.
I have performed a merge between the code you suggested and mine.